pinc/ProofProject.inc

```diff
@@ -0,0 +1,354 @@
+<?php
+declare(strict_types=1);
+
+trait ProofPage
+{
+    private function render_page_data(): array
+    {
+        $round_info_array = [];
+        foreach ($this->round->other_rounds_with_visible_usernames as $other_round_id) {
+            $other_round = get_Round_for_round_id($other_round_id);
+            $username = $this->get_username_for_round($other_round);
+            $round_info_array[] = [$other_round->id, $username];
+        }
+        $page_option = ($this->page_state == $this->round->page_out_state) ? "previous" : "current";
+        return
+            $this->get_page_text_array($page_option) +
+            $this->get_page_state_array() +
+            [
+                "pageNum" => pathinfo($this->page_name, PATHINFO_FILENAME),
+                "roundInfoArray" => $round_info_array,
+                "pagename" => $this->page_name,
+                "image_url" => "{$this->project->url}/{$this->page_name}",
+            ];
+    }
+
+    private function get_username_for_round(object $round): string
+    {
+        $sql = sprintf(
+            "
+            SELECT $round->user_column_name
+            FROM $this->projectid
+            WHERE image='%s'
+            ",
+            DPDatabase::escape($this->page_name)
+        );
+        $res = DPDatabase::query($sql);
+        $row = mysqli_fetch_assoc($res);
+        return $row[$round->user_column_name];
+    }
+
+    private function get_page_text_array(string $page_option): array
+    {
+        if ($page_option == "current") {
+            $desired_column_name = $this->round->text_column_name;
+        } else {
+            $desired_column_name = $this->round->prevtext_column_name;
+        }
+
+        $sql = sprintf(
+            "
+            SELECT $desired_column_name
+            FROM $this->projectid
+            WHERE image='%s'
+            ",
+            DPDatabase::escape($this->page_name)
+        );
+        $res = DPDatabase::query($sql);
+        $text = mysqli_fetch_row($res)[0];
+
+        return ["text" => $text];
+    }
+
+    private function get_page_state_array()
+    {
+        return [
+            "pagestate" => $this->page_state,
+            "saved" => ($this->page_state == $this->round->page_temp_state),
+        ];
+    }
+}
+
+class ProofProject
+{
+    use ProofPage;
+
+    public function __construct(object $project, string $state)
+    {
+        global $pguser;
+
+        if ($state != $project->state) {
+            $err = sprintf(
+                _('Warning: Project "%1$s" is not in state "%2$s"; it is now in state "%3$s".'),
+                $project->nameofwork,
+                project_states_text($state),
+                project_states_text($project->state)
+            );
+            throw new ProjectNotInRequiredStateException($err);
+        }
+
+        $project_round = $project->get_project_available_round();
+        $project_round->validate_user_can_access($pguser);
```

Comment:
Why not put all this validation in a validator?

Reply:
> Why not put all this validation in a validator?

If you are just suggesting putting this in a separate function somewhere else, what would be the point? If you were to define a new endpoint component that required this validation, such as `proofable_project` then you could create this ProofProject object in a validator. It's another way of doing it that would probably work just as well.